InvalidLinkBear: Reduce severity

satwikkansal opened the pull request:
Reduces the severity of the XML namespaces which are
present in xmlns attributes of the file. If the
same namespace link occurs more than once in the file,
it should not be reported.

Closes https://github.com/coala/coala-bears/issues/1239

pratyushprakash commented on tests/general/InvalidLinkBearTest.py:
```diff
@@ -128,7 +128,11 @@ def test_run(self):
+        http://example.co.in/404
+        <ruleset name="test" xmlns="http://this.isa.namespace/ruleset/7.0.0"
+        xmlns:xsi="http://this.is.another/kindof/namespace"
+        xsi:schemaLocation="http://this.namespace.dosent/exists/7.0.0"
```
IMO there should be a check for the severity

satwikkansal replied:
If that's the case, I guess we should create a separate issue for that because none of the tests are written yet to check the severity.

pratyushprakash replied:
It should be fairly simple. I don't think there is a need to create an issue

satwikkansal replied:
Tests added

jayvdb replied:
Done. Thanks.

pratyushprakash commented on tests/general/InvalidLinkBearTest.py:
```diff
@@ -128,7 +128,11 @@ def test_run(self):
+        <ruleset name="test" xmlns="http://this.isa.namespace/ruleset/7.0.0"
+        xmlns:xsi="http://this.is.another/kindof/namespace"
+        xsi:schemaLocation="http://this.namespace.dosent/exists/7.0.0"
+        xsi:schemaLocation="http://httpbin.com/404">"""
```
This is just ensuring there is 100% coverage. Since the PR deals with the severity of the result I feel it should be tested.

satwikkansal replied:
+1

jayvdb replied:
Resolved. Thanks.

satwikkansal commented:
I always forget to amend my commit after fixing PEP8 issues :(

pratyushprakash commented:
ack fcd02b1373fae90a7cf9131ceb07c2a418de52bd

jayvdb commented on bears/general/InvalidLinkBear.py:
```diff
@@ -80,19 +81,25 @@ def find_links_in_file(file, network_timeout, link_ignore_regex,
                                         # Unbalanced parenthesis
             (?<!\.)(?<!,)               # Exclude trailing `.` or `,` from URL
             """, re.VERBOSE)
+        xml_namespace_links = set()
```
I'm sure there will be other cases like this. Create an enum for 'first context' of each url, and start it with only 'xml namespace'.

satwikkansal replied:
:+1:

jayvdb commented on bears/general/InvalidLinkBear.py:
```diff
@@ -80,19 +81,25 @@ def find_links_in_file(file, network_timeout, link_ignore_regex,
                 if not (link_ignore_regex.search(link) or
-                        fnmatch(link, link_ignore_list)):
+                        fnmatch(link, link_ignore_list) or
+                        link in xml_namespace_links):
```
argh. This should be happening already for all links. See https://github.com/coala/coala-bears/issues/1319

We should actually report additional occurrences, but re-use the initial context of the url so that extra occurences are a lower severity.

satwikkansal replied:
Okay

jayvdb commented on bears/general/InvalidLinkBear.py:
```diff
@@ -80,19 +81,25 @@ def find_links_in_file(file, network_timeout, link_ignore_regex,
                                link, network_timeout.get(
                                    host, InvalidLinkBear.DEFAULT_TIMEOUT))
-                    yield line_number + 1, link, code
+                    yield line_number + 1, link, code, False
```
Return the 'context' enum in a dict, instead of a bool, so it is extensible.

satwikkansal replied:
Done

pratyushprakash commented:
unack fcd02b1373fae90a7cf9131ceb07c2a418de52bd

jayvdb commented on bears/general/InvalidLinkBear.py:
```diff
@@ -80,14 +82,22 @@ def find_links_in_file(file, network_timeout, link_ignore_regex,
                         fnmatch(link, link_ignore_list)):
                     if link.startswith(('hg+', 'bzr+', 'git+', 'svn+')):
-                        link = InvalidLinkBear.parse_pip_vcs_url(link)
+                        parsed_link = InvalidLinkBear.parse_pip_vcs_url(link)
```
just a note : this can become another context , e.g. "pip_vcs", in the future (not this PR).

satwikkansal replied:
I'll create an issue to identify other possible contexts and implement them.

sils commented:
So I do think we should simplify this code.  Also "Reduce Severity" doesn't really describe it, does it?

> If the
> satme namespace link occurs more than once in the file,
> it should not be reported.

that almost sounds like an own commit.

Can we maybe have like one function that takes all the lines and yields all the links with the protocol and their sourceposition and stuff? Maybe group links belonging together? I'm not fully sure but it seems like this could enefit from *some* simplification.

satwikkansal commented:
CircleCI is failing due to `cabal`

satwikkansal commented:
@sils I've simplified the code and split the changes into 2 commits, I guess it's ready to be merged after `CircleCI` stops displaying the strange `cabal` installation error :(

satwikkansal commented:
All tests are passing :smile:

jayvdb commented on bears/general/InvalidLinkBear.py:
```diff
@@ -80,22 +80,41 @@ def find_links_in_file(file, network_timeout, link_ignore_regex,
+                file, link_ignore_regex, link_ignore_list):
+
+            if link_context.pip_vcs_url:
+                parsed_link = InvalidLinkBear.parse_pip_vcs_url(link)
```
this should be `link = ...`

jayvdb commented on tests/general/InvalidLinkBearTest.py:
```diff
@@ -139,11 +161,41 @@ def test_run(self):
+        http://httpbin.org/status/503
+        http://httpbin.org/status/301
+        http://httpbin.org/status/302
+        xsi:schemaLocation="http://httpbin.com/404
```
a closing `"` would be helpful

satwikkansal commented:
@jayvdb Implemented the changes you suggested. I had a doubt btw, 
If we detect a url as an xml namespace and assign it a context of `xml_namespace`, Now shall the every occurrence of this particular url be yielded as low severity (even the ones which are not part of `xmlns:` or `xsi:` tags)?

jayvdb commented:
@satwikkansal, yes ... that context is global (whereas as currently your pip vcs context is not global - that can be fixed).  If a link has been found to be a schema, that automatically says the link may be valid even if it is a 404 - the link doesnt need to have http content for its *intended* use.  It is a declared namespace - that is all the URI is intended to be.  If it is mentioned anywhere in a README or any technical documentation, it is still valid.  This presents a problem for the per-file non-global design of this bear, as it means files processed before the xml namespace is seen will be errors while those after are not.  We can defer that problem to a separate issue, if you like.  The main problem is removing the inter-file errors.

satwikkansal commented:
@jayvdb Rebased. I guess it's ready to be merged.

Regarding the global context, IMO, we can have a preprocessing stage where the  bear only populates the context of all the links across the files it is run on, and after that, check the status and then generate results based on the status and the context.
Shall this be a part of https://github.com/coala/coala-bears/issues/1319 of shall I create a separate issue for this.

jayvdb commented on tests/general/InvalidLinkBearTest.py:
```diff
@@ -86,6 +87,24 @@ def assertResultCount(self, test_file, expected_num):
+        """
+        severity_tag = RESULT_SEVERITY.reverse[severity]
+        with requests_mock.Mocker() as m:
+            InvalidLinkBear.check_prerequisites = lambda *args: True
```
I think this should be removed.
I assume it is for performance reasons.  Create another bug about that ;-)

satwikkansal replied:
Do you want me to remove complete `assertSeverity` function or just the single line.

jayvdb commented on tests/general/InvalidLinkBearTest.py:
```diff
@@ -139,11 +166,44 @@ def test_run(self):
+        info_severity_file = '''
+        <ruleset name="test" xmlns="http://this.isa.namespace/ruleset/7.0.0"
+        xmlns:xsi="http://this.is.another/kindof/namespace">
+        <ruleset name="test" xmlns="http://this.isa.namespace/ruleset/7.0.0"
```
As the parser is context aware, using the same URL multiple times in one test sample input is not as effective as using different URLs.  I think the URL in this `<ruleset>` should be slightly different - maybe different number, or slightly different domain name.

jayvdb commented on tests/general/InvalidLinkBearTest.py:
```diff
@@ -139,11 +166,44 @@ def test_run(self):
+    def test_severity(self):
+        info_severity_file = '''
+        <ruleset name="test" xmlns="http://this.isa.namespace/ruleset/7.0.0"
+        xmlns:xsi="http://this.is.another/kindof/namespace">
```
please close the XML tag using `/>`.
the fragment will still not be a valid document, but at least the fragment is not invalid XML.

jayvdb commented on tests/general/InvalidLinkBearTest.py:
```diff
@@ -139,11 +166,44 @@ def test_run(self):
+        self.assertSeverity(normal_severity_file, RESULT_SEVERITY.NORMAL)
+
+        major_severity_file = '''
+        http://coalaisthebest.com
```
Using `http://this.isa.namespace/` , and putting it at the top of this test method, would help the reader understand.

satwikkansal commented:
@jayvdb Addressed your suggestions. Also, shifted the severity tests to a separate commit as they were getting confusing.

> Also can we have a test case which is exactly the scenario in the issue being fixed, so we dont have a regression somehow.

Done. Added all the relevant test cases in `test_xml_namespaces` method.

> Yes, global scope should be a separate issue.

Okay. I'll create an issue for global scope once this commit related to link contexts https://github.com/coala/coala-bears/pull/1297/commits/54eeec270e31cb0330a082f72494b21aa08b0f51 gets merged.

jayvdb commented:
ack a5a2fb3

jayvdb commented:
ack bbd5d33 7f99537 9306520

jayvdb commented:
@rultor merge

rultor commented:
> @rultor merge

@jayvdb OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/11330-291836877)

rultor commented:
> @rultor merge

@jayvdb Done! FYI, the full log is [here](http://www.rultor.com/t/11330-291836877) (took me 1min)
